src/transport/stream/transport.rs

```diff
@@ -180,16 +181,10 @@ where
 		});
 
 		// Keep writing the header until it is done.
-		while this.bytes_written < FRAMED_HEADER_LEN {
+		while this.bytes_written < FRAMED_HEADER_LEN + body.len() {
 			let stream = Pin::new(&mut this.stream);
-			this.bytes_written += ready!(stream.poll_write(context, &header_buffer[this.bytes_written..]))?;
-		}
-
-		// Keep writing the body contents until it is done.
-		while this.bytes_written - FRAMED_HEADER_LEN < body.len() {
-			let body_written = this.bytes_written - FRAMED_HEADER_LEN;
-			let stream = Pin::new(&mut this.stream);
-			this.bytes_written += ready!(stream.poll_write(context, &body.data[body_written..]))?;
+			let bufs: &[_] = &[IoSlice::new(&header_buffer[..]), IoSlice::new(&body.data)];
```

Comment:
You need to remember how much is written and remove the leading data when creating the `IoSlice` array. If the entire header has already been written, you even need to remove the entire `IoSlice` for the header.

Reply:
Is there any way that we could unit test that behavior?

Reply:
I think the only way to trigger this is to send a large enough body so that the kernel will indeed split the writes into multiple syscalls. It's hard to get a guarantee on that though.

Reply:
On Unix, we could use a socket on which we've set the `SO_SNDBUF` option. From `man 7 socket`:

>        SO_SNDBUF
>              Sets  or  gets the maximum socket send buffer in bytes.  The kernel doubles this value (to allow space
>              for bookkeeping overhead) when it is set using setsockopt(2), and this doubled value  is  returned  by
>              getsockopt(2).   The  default value is set by the /proc/sys/net/core/wmem_default file and the maximum
>              allowed value is set by the /proc/sys/net/core/wmem_max file.  The minimum (doubled)  value  for  this
>              option is 2048.

If we set it to the minimum and send a larger body, we should cause the kernel to perform a partial write.